### PR TITLE
build: improve opt deps

### PR DESCRIPTION
**Pull request by nstarman**
Small improvements to the specification of optional dependencies. In particular, this improves `astropy[all]` to actually have everything and installs pre-commit for `astropy[test]` since our CI consists of 1. pre-commit checks, e.g. `yaml` verification, and 2. `pytest`-powered stuff.

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
Looks reasonable but let me run the other jobs, just in case. Thanks!

**Comment by astrofrog** [comment]
Should all actually include test_all?

**Comment by nstarman** [comment]
As in I should reverse the relationship and have `all` be a subset of `test_all`, not `test_all` a subset of `all` ?

all = [...]  # doesn't have test_all
test_all = ["astropy[all]", ...]

**Comment by pllim** [comment]
> have `all` be a subset of `test_all`

I think this makes more sense. What do you think, @astrofrog ?

File: pyproject.toml [hunk 1]
```diff
@@ -72,8 +73,12 @@ typing = [
     "typing_extensions>=4.0.0"
 ]
 all = [
-    "astropy[recommended]",  # installs the [recommended] dependencies
+    # Install grouped optional dependencies
+    "astropy[recommended]",
+    "astropy[docs]",
```
Comment (eerovaher) [inline pyproject.toml:78]: `test_all` has already been discussed, but I don't think `all` dependencies should contain `docs` dependencies either.

**Comment by nstarman** [comment]
It would be nice to have an "absolutely everything" category, which makes getting setting up astropy dev environments super simple. I was making `all` into that, but it's AOK to keep that as all runtime dependencies. What do people think about a `dev` category that contains all dependencies  -- runtime, static, docs, testing ?

**Comment by mhvk** [comment]
Agreed that I always thought of `all` as getting meaning I can now run absolutely every astropy function, not that it includes testing or building documentation. Hence, `all` being part of `test_all` makes more sense to me, just like `astropy[recomended]` is part of `docs`.

I like the idea of a `dev` dependency that includes all needed to start development, but slightly orthogonal here. E.g., not quite clear that needs all of astropy's optional dependencies (I guess that would be `dev_all`...)

File: pyproject.toml [hunk 2]
```diff
@@ -88,11 +91,8 @@ all = [
     "mpmath",
     "asdf-astropy>=0.3",
     "bottleneck",
-    "ipython>=4.2",
```
Comment (mhvk) [inline pyproject.toml:91]: Actually, given our discussion over in #16351, perhaps we shouldn't have "hidden" dependencies - code has a dependence on whether `ipython` is installed in `utils.console`.

But possibly that is best done separately - here, what we list is the dependencies needed for all astropy functionality.

Anyway, am happy either with this as-is, or with `ipython` put back.

**Review by mhvk — APPROVED** [review]
See my one remaining comment, which mostly means another reason to leave this open for a little longer so others can comment.

But I think this little reorganization makes good sense now!

File: pyproject.toml (hunk 2 above)
Comment (neutrinoceros) [reply pyproject.toml:91]: #16354 👀

Comment (eerovaher) [reply pyproject.toml:91]: Any dependency we have should be listed exactly once, but it would indeed be better to keep `ipython` here and to delete the duplicate from the `test_all` set instead.

File: pyproject.toml [hunk 3]
```diff
@@ -56,6 +57,7 @@ test = [
     "threadpoolctl",
 ]
 test_all = [
+    "astropy[all]",  # installs all optional run-time dependencies
     "astropy[test]",  # installs the [test] dependencies
```
Comment (eerovaher) [inline pyproject.toml:61]: The first comment is useful (because it mentiones "run-time"), the second comment is not. The useless comment isn't added by this pull request, but several analogous and equally useless comments are being introduced that should all be cleaned up.
```suggestion
    "astropy[all]",  # installs all optional run-time dependencies
    "astropy[test]"
```

File: pyproject.toml [hunk 4]
```diff
@@ -105,6 +105,16 @@ docs = [
     "tomli; python_version < '3.11'",
     "sphinxcontrib-globalsubs >= 0.1.1",
 ]
+dev = [
+    "astropy[recommended]",  # installs the most common optional dependencies
+    "astropy[test]",  # installs the [test] dependencies
+    "astropy[docs]",  # installs the [docs] dependencies
+    "astropy[typing]",  # installs the [typing] dependencies
+]
+dev_all = [
```
Comment (eerovaher) [inline pyproject.toml:114]: These new installation options should be (briefly) introduced somewhere in the developer documentation.

**Comment by nstarman** [comment]
Given the merge conflict I took this opportunity to add code comments and some discussion in the docs. I also added the `[ipython]` and `[utils.console]` optional dependency categories to group the newly added dependencies (the source of the merge conflict). Lastly, I put the run-time groups at the top.

File: pyproject.toml (hunk 4 above)
Comment (nstarman) [reply pyproject.toml]: Done.

File: pyproject.toml (hunk 2 above)
Comment (nstarman) [reply pyproject.toml]: Done.

File: pyproject.toml (hunk 3 above)
Comment (nstarman) [reply pyproject.toml]: Done.

File: pyproject.toml (hunk 4 above)
Comment (nstarman) [reply pyproject.toml]: Done.

File: pyproject.toml [hunk 5]
```diff
@@ -90,13 +85,32 @@ all = [
...
+    "pytest-xdist",
+    "threadpoolctl",
+]
+test_all = [
+    # Install grouped optional dependencies
+    "astropy[all]",  # installs all optional run-time dependencies
+    "astropy[test]",
+    # Install all remaining dependencies
+    "objgraph",
+    "coverage[toml]",
```
Comment (neutrinoceros) [inline pyproject.toml:107]: shouldn't this line be in `[test]` ? I also note that it's actually already there in a way (`pytest-astropy` requires `pytest-cov` which itself requires `coverage[toml]`)

File: pyproject.toml [hunk 6]
```diff
@@ -90,13 +85,32 @@ all = [
...
+    "skyfield>=1.20",
+    "sgp4>=2.3",
+    "array-api-strict",
+]
+typing = [
+    "typing_extensions>=4.0.0"
```
Comment (neutrinoceros) [inline pyproject.toml:113]:
```suggestion
    "typing_extensions>=4.0.0",
```

File: docs/install.rst [hunk 7]
```diff
@@ -437,7 +437,7 @@ installing ``astropy`` with pip::
 You can alternatively install the `sphinx-astropy
 <https://github.com/astropy/sphinx-astropy>`_ package with pip::
 
-    python -m pip install sphinx-astropy
+    python -m  sphinx-astropy
```
Comment (neutrinoceros) [inline docs/install.rst:440]:
```suggestion
    python -m sphinx-astropy
```

**Review by neutrinoceros — APPROVED** [review]
Just a couple of extremely minor suggestions

**Review by eerovaher — COMMENTED** [review]
I'm not convinced we need separate `ipython` and `jupyter` dependency sets.

It does look like it would be a good idea to rewrite the `astropy` development version installation instructions and the dependency descriptions, but doing that properly is beyond the scope of this pull request.

File: docs/install.rst (hunk 7 above)
Comment (neutrinoceros) [reply docs/install.rst:440]: good point, I may have been low on coffee if I didn't see this. Let me update my suggestion.

File: pyproject.toml [hunk 8]
```diff
@@ -47,35 +47,30 @@ dependencies = [
...
+# you are using Astropy from an IPython-dependent IDE, like Jupyter, this should enforce
+# the minimum supported version of IPython.
+ipython = [
+    "ipython>=7.32",
+]
+jupyter = [  # these are optional dependencies for `utils.console`
+    "astropy[ipython]",
+    "ipywidgets",
+    "ipykernel",
 ]
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply pyproject.toml:66]: I don't think it does. But I'm also not sure it should: `jupyter` comes in different shapes (`jupyterlab` and `notebook`) and we don't need to impose a choice onto our users (I think ?)

[the remark this answers is not in the document]
Comment (pllim) [reply pyproject.toml:66]: Something seems to be pulling in jupyter stack but it is already that way in main. Hmm!

```
2024-05-19T22:19:37.9405799Z jupyter_client==8.6.1
2024-05-19T22:19:37.9406161Z jupyter_core==5.7.2
2024-05-19T22:19:37.9406568Z jupyterlab_widgets==3.0.10
```

File: docs/install.rst (hunk 7 above)
Comment (nstarman) [reply docs/install.rst:440]: Shouldn't have been. I think this happened in the rebase. I'll edit.

**Review by mhvk — COMMENTED** [review]
Nice! Good to go with the remaining suggested changes from others as far as I'm concerned.

File: pyproject.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (nstarman) [reply pyproject.toml:66]: This is just moving around the dependencies, not adding any. I titled it jupyter since it's for use when in a Jupyter environment.

File: docs/install.rst (hunk 7 above)
Comment (nstarman) [reply docs/install.rst]: Done.

File: docs/development/codeguide.rst [hunk 9]
```diff
@@ -37,8 +37,7 @@ Interface and Dependencies
   statement, which will raise an ``ImportError`` if the dependency is
   not available. In the astropy core package, such optional dependencies should
   be recorded in the ``pyproject.toml`` file in the ``[project.optional-dependencies]``
-  entry, under ``all`` (or ``test_all`` if the dependency is only
-  needed for testing).
+  entry (try to put it in the appropriate category of optional dependencies).
```
[the remark this answers is not in the document]
Comment (nstarman) [reply docs/development/codeguide.rst]: Done.

File: docs/development/testguide.rst [hunk 10]
```diff
@@ -39,8 +39,17 @@ option::
 
     python -m pip install --editable ".[test]"
 
-A detailed description of the plugins can be found in the :ref:`pytest-plugins`
-section.
+To test the full set of optional dependencies, use the ``test_all`` option::
+
+    python -m pip install --editable ".[test_all]"
+
+If you are looking to do development in Astropy beyond running the tests, e.g. building the documentation or doing static analysis, we provide the complete set of all
```
[the remark this answers is not in the document]
Comment (nstarman) [reply docs/development/testguide.rst]: Fixed.

File: pyproject.toml (hunk 5 above)
Comment (nstarman) [reply pyproject.toml]: Moved!

File: pyproject.toml (hunk 6 above)
Comment (nstarman) [reply pyproject.toml]: Added.

**Comment by pllim** [comment]
> with the remaining suggested changes from others

Does this mean we're still seeking a couple of more approvals?

**Review by eerovaher — COMMENTED** [review]
I'm still not convinced we need separate `ipython` and `jupyter` dependency sets, but listing the dependencies is definitely done much better in this pull request than on `main`.

**Comment by mhvk** [comment]
> > with the remaining suggested changes from others
> 
> Does this mean we're still seeking a couple of more approvals?

I thought the comments given by others had not yet been all addressed, so was hoping to see those addressed or dismissed.

File: pyproject.toml (hunk 8 above)
[the remark this answers is not in the document]
Comment (nstarman) [reply pyproject.toml:66]: This is in contrast to the ipython dep, which enables functionality in utils.console within any ipython environment not just a Jupyter one.

[the remark this answers is not in the document]
Comment (eerovaher) [reply pyproject.toml:66]: Do you expect anyone to ever use the `ipython` and `jupyter` dependency sets?

[the remark this answers is not in the document]
Comment (neutrinoceros) [reply pyproject.toml:66]: Yes, AFAIC. Even if not publicly advertised, these things have a way of finding their users, in my experience.

[the remark this answers is not in the document]
Comment (nstarman) [reply pyproject.toml:66]: For another PR, but I do think we should have a page detailing these install options. Like...

If you're using an IPYthon environment, use `astropy[ipython]`. This is further improved by `astropy[jupyter]` when in a Jupyter notebook. `astropy[typing]` improves static typing compatibility. To get everything do `astropy[all]`.

**Comment by mhvk** [comment]
Looking at this again, I think there are actually no outstanding issues - and if there are, these can be dealt with later. So, let's get this in. Thanks, @nstarman, this is substantially nicer!

**Comment by pllim** [comment]
tl;dr -- Did you double check that `tox.ini` and the workflows in `.github/workflows` not need updating?

File: pyproject.toml [hunk 11]
```diff
@@ -90,13 +85,32 @@ all = [
     "mpmath",
     "asdf-astropy>=0.3",
     "bottleneck",
-    "ipykernel",
-    "ipython>=7.32",
-    "ipywidgets",
-    "pytest>=7.0",
```
Comment (bsipocz) [inline pyproject.toml:96]: Why has been pytest removed as a dependency here? The test runner uses it, so strictly speaking it's a genuine optional dependency.
